File: src/framework/controllers/auth/sign-out.ts
import {
  AuthenticatedLocals,
  IRequest,
  IResponse,
} from "../../../common/types";
import { z } from "zod";
import { Controller, HttpMethod } from "../types";
import databases from "../../../data/data-sources/injection";
import { authMiddleware } from "../../middleware";
import { getSocketServer } from "../../adapters/injection";

export class SignOutController implements Controller<void> {
  method = HttpMethod.POST;

  route = "/auth/signOut";

  validator = z.object({
    allDevices: z.boolean(),
  });

  middleware = [authMiddleware];

  execute(
    req: IRequest<z.infer<typeof this.validator>>,
    res: IResponse<void, AuthenticatedLocals>
  ) {
    getSocketServer().kickUser(res.locals.cognitoId);
    return databases().authDatabase.signOut(req.authorization);
  }
}

File: lib/src/data/backend/auth/sign_out.dart
part of backend;

extension SignOut on Auth {
  Future<Failure?> signOut({
    required String accessToken,
    required bool allDevices,
  }) =>
      _request(
        route: "/auth/signOut",
        method: HttpMethod.post,
        data: {
          'allDevices': allDevices,
        },
        authorization: accessToken,
      );
}
